core/src/main/scala/akka/kafka/internal/ConsumerStage.scala

```diff
@@ -0,0 +1,267 @@
+    }
+
+    override def commit(): Future[Done] = {
+      if (offsets.isEmpty)
+        Future.successful(Done)
+      else {
+        stage match {
+          case Some(s) => s.commit(offsets)
+          case None =>
+            Future.failed(new IllegalStateException(s"commit not supported, $offsets"))
+        }
+      }
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[kafka] class ConsumerStage[K, V](consumerProvider: () => KafkaConsumer[K, V])
+    extends GraphStageWithMaterializedValue[SourceShape[Consumer.CommittableMessage[K, V]], Consumer.Control]
+    with LazyLogging {
+  import ConsumerStage._
+
+  val out = Outlet[Consumer.CommittableMessage[K, V]]("messages")
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val consumer = consumerProvider()
+    var stopping = false
+    val stoppedPromise = Promise[Done]
+
+    val logic = new TimerGraphStageLogic(shape) with Committer with Consumer.Control {
+      private case object Poll
+      // FIXME move out settings
+      private val POLL_INTERVAL = 100.millis
+      private val POLL_DATA_TIMEOUT = 50.millis
+      private val POLL_COMMIT_TIMEOUT = 1.millis
+      private val COMMIT_TIMEOUT = 10.seconds
+      private var awaitingConfirmation = 0L
+      private var buffer: Iterator[Consumer.CommittableMessage[K, V]] = Iterator.empty
+      private var pollScheduled = false
+
+      private def schedulePoll() = {
+        if (!pollScheduled) {
+          pollScheduled = true
+          scheduleOnce(Poll, POLL_INTERVAL)
+        }
+      }
+
+      private def poll() = {
+        def setupConsumer() = {
+          // TODO do we really need to support change of assignments after startup?
+          //      Lots of allocations here.
+          if (isAvailable(out)) {
+            consumer.resume(consumer.assignment().asScala.toSeq: _*)
+          }
+          else {
+            consumer.pause(consumer.assignment().asScala.toSeq: _*)
+          }
+        }
+
+        def handleResult(records: ConsumerRecords[K, V]) = {
+          if (!records.isEmpty) {
+            logger.trace(s"Got messages - {}", records)
+            require(!buffer.hasNext)
+            require(isAvailable(out))
+            buffer = records.iterator().asScala.map { record =>
+              val offset = Consumer.Offset(
+                topic = record.topic,
+                partition = record.partition,
+                offset = record.offset
+              )
+              val committable = CommittableOffsetImpl(offset)(this)
+              Consumer.CommittableMessage(record.key, record.value, committable)
+            }
+            pushMsg(buffer.next())
+          }
+        }
+
+        def pollTimeout = {
+          val data = Option(POLL_DATA_TIMEOUT).filter(_ => isAvailable(out) && !buffer.hasNext)
+          def commit = Option(POLL_COMMIT_TIMEOUT).filter(_ => awaitingConfirmation > 0)
+          data.orElse(commit)
```

Comment:
Could be implemented with much fewer allocations (especially the `Option(CONSTANT)`s), if folded with the pollTimeout pattern match it could probably be allocation free even?